Add Raft TLS Helm examples (#12982)
hashicorp/vault: website/content/docs/platform/k8s/helm/examples/ha-tls.mdx

---
layout: 'docs'
page_title: 'HA Cluster with TLS'
sidebar_current: 'docs-platform-k8s-examples-ha-tls'
description: |-
  Describes how to set up a Raft HA Vault cluster with TLS certificate
---

> # Raft HA Server with TLS 

Reviewer: ```suggestion
# Raft HA server with TLS
```

Style correction: use sentence case for headings

> Follow the steps from the example [HA Vault Cluster with Integrated Storage](/docs/platform/k8s/helm/examples/ha-with-raft) to build the cluster.
> 
> Follow the examples and instructions in [Standalone Server with TLS](/docs/platform/k8s/helm/examples/standalone-tls) to create a TLS certificate.

Reviewer: ```suggestion

## Before you start

1. Follow the steps in [HA Vault Cluster with Integrated Storage](/docs/platform/k8s/helm/examples/ha-with-raft) to build your cluster.
1. Follow the steps in [Standalone Server with TLS](/docs/platform/k8s/helm/examples/standalone-tls) to create a TLS certificate.
```

Suggest making it clear that these are (essentially) prerequisites to following the instructions you provide below.

> 

Reviewer: ```suggestion
## Bootstrap your Raft cluster

```

Suggest adding a heading to make it obvious what the suggestions below will have folks do

> Before cluster initialization and without proper configuration, the following warning is to be expected:

Reviewer: ```suggestion
Without proper configuration, you will see the following warning before cluster initialization:
```

style correction: write in active voice

```shell
core: join attempt failed: error="error during raft bootstrap init call: Put "https://vault-${N}.${SERVICE}:8200/v1/sys/storage/raft/bootstrap/challenge": x509: certificate is valid for ${SERVICE}, ${SERVICE}.${NAMESPACE}, ${SERVICE}.${NAMESPACE}.svc, ${SERVICE}.${NAMESPACE}.svc.cluster.local, not vault-${N}.${SERVICE}"
```

> The concepts for [Integrated Storage and TLS](/docs/concepts/integrated-storage#integrated-storage-and-tls) elaborate three possible solutions to mitigate this TLS verification warning and bootstrap the Raft cluster, two of which are discussed in further detail here.

Reviewer: ```suggestion
The concepts overview for [Integrated Storage and TLS](/docs/concepts/integrated-storage#integrated-storage-and-tls) covers the various options for mitigating TLS verification warnings and bootstraping your Raft cluster. 

The examples below demonstrate two specific solutions. Both solutions ensure that the common name (CN) used for the `leader_api_addr` in the Raft stanza matches the name(s) listed in the TLS certificate.
```
I suggest moving this up to the beginning of the document (before the suggested "Before you start" section) as it's a useful introduction to what your document will cover. Also, I would add the last sentence from your doc here as well since it provides a good summary of what your specific examples do.

Style correction: write in active voice.

> The warning disappears when the the [expected TLS servername](/docs/concepts/integrated-storage#autojoin-with-tls-servername) is correctly configured using [`leader_tls_servername`](/docs/configuration/storage/raft#leader_tls_servername) in the Raft stanza (`${CN}` in the example below):

Reviewer: ```suggestion

## Solution 1: Use auto-join and set the TLS server in your Raft configuration

The join warning disappears if you use autojoin and set the expected TLS server name (`${CN}`) with [`leader_tls_servername`](/docs/configuration/storage/raft#leader_tls_servername) in the Raft stanza for your Vault configuration.

For example:
```

Suggest making it easier for folks scanning the doc to differentiate between the two solutions you're demonstrating.

Style correction: write in active voice

> ```hcl

Reviewer: It might be useful to highlight the important line for folks so they know which part of the Raft config you're talking about. You can highlight the relevant line by wrapping the `CodeBlockConfig` component around your code example (with whitespace before/after each tag) and noting the specific line(s) you want to highlight. For example:

```
<CodeBlockConfig highlight="6,14,22">

YOUR_CODE_BLOCK

</CodeBlockConfig>
```

storage "raft" {
  path = "/vault/data"

  retry_join {
    leader_api_addr = "https://vault-0.${SERVICE}:8200"
    leader_tls_servername = "${CN}"
    leader_client_cert_file = "/vault/tls/vault.crt"
    leader_client_key_file = "/vault/tls/vault.key"
    leader_ca_cert_file = "/vault/tls/vault.ca"
  }

  retry_join {
    leader_api_addr = "https://vault-1.${SERVICE}:8200"
    leader_tls_servername = "${CN}"
    leader_client_cert_file = "/vault/tls/vault.crt"
    leader_client_key_file = "/vault/tls/vault.key"
    leader_ca_cert_file = "/vault/tls/vault.ca"
  }

  retry_join {
    leader_api_addr = "https://vault-2.${SERVICE}:8200"
    leader_tls_servername = "${CN}"
    leader_client_cert_file = "/vault/tls/vault.crt"
    leader_client_key_file = "/vault/tls/vault.key"
    leader_ca_cert_file = "/vault/tls/vault.ca"
  }
}
```

> Alternatively, specify one `retry_join` which references the [name of the load balancer](/docs/concepts/integrated-storage#load-balancer-instead-of-autojoin) instead:

Reviewer: ```suggestion

## Solution 2:  Add a load balancer to your Raft configuration

If you have a load balancer for your Vault cluster, you can add  a single `retry_join` stanza to your Raft configuraiton and use the load balancer address for `leader_api_addr`.

For example:
```

Same suggestion as above: make it easier on folks scanning quickly to find the start of your second recommended solution

> ```hcl

Reviewer: If you want to highlight the relevant line for this example:

```
<CodeBlockConfig highlight="5">

YOUR_CODE_BLOCK

</CodeBlockConfig>
```

storage "raft" {
  path = "/vault/data"

  retry_join {
    leader_api_addr = "https://vault-active:8200"
    leader_client_cert_file = "/vault/tls/vault.crt"
    leader_client_key_file = "/vault/tls/vault.key"
    leader_ca_cert_file = "/vault/tls/vault.ca"
  }
}
```

> Both configuration options ensure that the the common name (CN) used for the `leader_api_addr` in the Raft stanza matches the name(s) listed in the TLS certificate.

Reviewer: ```suggestion
```

Suggest moving this statement to earlier in the doc (I added it to a previous suggestion).
